src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -425,16 +426,24 @@ private static void UpdateSessionData(IAsyncCompletionSession session, Completio
             // If there are suggestionItemOptions, then later HandleNormalFiltering should set selection to SoftSelection.
             sessionData.HasSuggestionItemOptions |= completionList.SuggestionModeItem != null;
 
-            var excludedCommitCharacters = GetExcludedCommitCharacters(completionList.ItemsList);
-            if (excludedCommitCharacters.Length > 0)
+            var excludedCommitCharactersFromList = GetExcludedCommitCharacters(completionList.ItemsList);
+            if (session.Properties.TryGetProperty(ExcludedCommitCharactersMap, out MultiDictionary<char, RoslynCompletionItem> excludedCommitCharactersMap))
             {
-                if (session.Properties.TryGetProperty(ExcludedCommitCharacters, out ImmutableArray<char> excludedCommitCharactersBefore))
+                foreach (var kvp in excludedCommitCharactersFromList)
```

Comment:
>foreach (var kvp in excludedCommitCharactersFromList) [](http://example.com/codeflow?start=16&length=53)

nit: would be nice if there were an AddRange on MultiDictionary. Would simplify this code and it's implementation would be slightly more performant as it wouldn't need to look up each key once

Reply:
I think `AddRange` is usually associated with list like types in BCL. Maybe we could add a `Union` for `MultiDictionary`? I can give this a try in a follow up PR since it'd require change in compiler layer.

Reply:
Sounds good, I just saw that OrderedMultiDictionary has an AddRange, so I thought the name could match that, but Union is fine by me too.

Reply:
Out of curiosity, this method must get called on the same session multiple times with different completionList.ItemsList values, right? Just trying to understand if there is a relation between previous values in the ExcludedCommitCharactersMap property and the items returned from GetExcludedCommitCharacters.

Reply:
Yes, currently it could be called twice for a given session to support import completion, as we might delay those unimported item to a second batch w/o showing them until the list refreshed, or if import completion is disabled, it can be explicitly triggered by using the expander button during the same session. Which means the values in the map from before vs after are from two disjoint sets of completion items (unimported item vs everything else).